nilearn/utils/discovery.py

```diff
@@ -0,0 +1,188 @@
+"""Utilities to discover nilearn objects."""
+
+import inspect
+import pkgutil
+from importlib import import_module
+from operator import itemgetter
+from pathlib import Path
+
+from sklearn.base import BaseEstimator, ClusterMixin, TransformerMixin
+from sklearn.utils._testing import ignore_warnings
+
+from nilearn._utils.param_validation import check_parameter_in_allowed
+
+_MODULE_TO_IGNORE = {
+    "_utils",
+    "conftest",
+    "externals",
+    "input_data",
+    "tests",
+}
+
+
+def _skip_module(module_name: str):
+    module_parts = module_name.split(".")
+    return bool(
+        any(part in _MODULE_TO_IGNORE for part in module_parts)
+        or "._" in module_name
+    )
+
+
+def all_estimators(type_filter=None):
+    """Get a list of all estimators from `nilearn`.
+
+    This function crawls the module and gets all classes that inherit
+    from sklearn BaseEstimator.
+    Classes that are defined in test-modules are not included.
+
+    Parameters
+    ----------
+    type_filter : {"classifier",  "cluster", "regressor", "masker", \
+                  "multi_masker", "transformer"} \
+                  or list of such strings, default=None
+        Which kind of estimators should be returned.
+        If ``None``, no filter is applied and all estimators are returned.
+        Possible values are
+        "classifier", "cluster", "regressor", "masker",
+        "multi_masker", "transformer"
+        to get estimators only of these specific types,
+        or a list of these to get the estimators
+        that fit at least one of the types.
+
+    Returns
+    -------
+    estimators : functions : :obj:`list` of :obj:`tuple`
+        List of (name, class),
+        where ``name`` is the class name as string
+        and ``class`` is the actual type of the class.
+
+    """
+    # lazy import to avoid circular imports
+    from nilearn.decoding._mixin import _ClassifierMixin, _RegressorMixin
+    from nilearn.maskers._mixin import _MultiMixin
+    from nilearn.maskers.base_masker import BaseMasker, _BaseSurfaceMasker
```

Comment:
Do we have dependency from these packages to `nilearn.utils` ? I think there should not be.
 
We are introducing `nilearn.utils` package now as a public package. I think that would be good to define what it serves and its boundaries in `__init__.py` of the package. What I understand is its aim is to provide users with some useful functionality.

In my opinion, this package should be able to import from other packages, but no other package should import from this package. (At least, in a way that might create cyclic dependency. However in any case, I think we should avoid this)